File: 20_valid_parenthesis/checking_brackets.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
       match = {
           ')':'(',
           '}':'{',
           ']':'['
       }
       opened = []
       for i in range(len(s)):
            if match.get(s[i], False):
               if len(opened)==0 or opened.pop() != match[s[i]]:
                   return False
            else:
                opened.append(s[i])
        return len(opened)==0

File: 20_valid_parenthesis/checking_brackets.cpp
#include<iostream>
#include<unordered_map>
#include<stack>
#include<cassert>

using namespace std;

class Solution{
public:
     bool isValid(string s) {
        std::unordered_map<char, char> match = {
            {')','('},
            {'}','{'},
            {']','['}
        };
        std::stack<char> opened;
        for(char ch: s) {
            if (match.find(ch) != match.end()) {
                if ( opened.empty() || opened.top() != match[ch]) {
                    return false;
                }
                else {
                    opened.pop();
                } 
            }
            else {
                opened.push(ch);
            }
        }
        return opened.empty() ;
    }
};
